File: android/app/src/main/java/com/rnbridge/HelloFromTheOtherSide.java
package com.rnbridge;

import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeArray;
import com.facebook.react.bridge.WritableNativeMap;

public class HelloFromTheOtherSide extends ReactContextBaseJavaModule {
    HelloFromTheOtherSide(ReactApplicationContext context) {
        super(context);
    }

    @Override
    public String getName() {
        return "HelloFromTheOtherSide";
    }

    @ReactMethod
    public void greet(String name, Promise promise) {
        promise.resolve("Hello " + name);
    }

    @ReactMethod
    public void generateNames(Integer numberOfNames, Promise promise) {
        String name = "Justin";
        WritableArray result = new WritableNativeArray();
        for (int i = 0; i < numberOfNames; i++) {
            result.pushString(name);
        }
        promise.resolve(result);
    }

    @ReactMethod
    public void capitaliseUserNames(ReadableMap user, Promise promise) {
        String firstNameKey = "firstName";
        String lastNameKey = "lastName";

        if (!user.hasKey(firstNameKey) || !user.hasKey(lastNameKey)) {
            promise.reject("missingKeys", "firstName and lastName must be present on user");
            return;
        }

        WritableMap result = new WritableNativeMap();
        result.putString(firstNameKey, user.getString(firstNameKey).toUpperCase());
        result.putString(lastNameKey, user.getString(lastNameKey).toUpperCase());
        promise.resolve(result);
    }
}

File: ios/rnbridge/HelloFromTheOtherSide.m
#import "HelloFromTheOtherSide.h"

@implementation HelloFromTheOtherSide

RCT_EXPORT_MODULE();

RCT_EXPORT_METHOD(greet:(NSString*)name
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  NSString *hello = @"Hello ";
  NSString *result = [hello stringByAppendingString:name];
  resolve(result);
};

RCT_EXPORT_METHOD(generateNames:(NSInteger)numberOfNames
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
  NSString *name = @"Justin";
  NSMutableArray *array = [[NSMutableArray alloc] init];
  for (int i = 0; i < numberOfNames; i++) {
    [array addObject:name];
  }
  NSArray *result = [array copy];
  resolve(result);
}

RCT_EXPORT_METHOD(capitaliseUserNames:(NSDictionary*)user
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
{
  NSMutableDictionary *mutableUser = [user mutableCopy];
  NSString *firstNameKey = @"firstName";
  NSString *lastNameKey = @"lastName";
  
  if ([mutableUser valueForKey:firstNameKey] == nil || [mutableUser valueForKey:lastNameKey] == nil) {
    reject(@"missingKeys", @"firstName and lastName must be present on user", nil);
    return; // otherwise it will reject AND resolve
  }
  
  mutableUser[firstNameKey] = [[NSString stringWithString:[NSString stringWithFormat:@"%@",[mutableUser objectForKey:firstNameKey]]] uppercaseString];
  mutableUser[lastNameKey] = [[NSString stringWithString:[NSString stringWithFormat:@"%@",[mutableUser objectForKey:lastNameKey]]] uppercaseString];
  resolve([mutableUser copy]);
}
@end
